packages/sitecore-jss-nextjs/src/edge/personalize-middleware.ts

```diff
@@ -0,0 +1,184 @@
+
+export type PersonalizeMiddlewareConfig = {
+  /**
+   * Function used to determine if route should be excluded from personalization.
+   * By default, files (pathname.includes('.')) and API routes (pathname.startsWith('/api')) are ignored.
+   * This is an important performance consideration since Next.js Edge middleware runs on every request.
+   * @param {string} pathname The pathname
+   * @returns {boolean} Whether to exclude the route from personalization
+   */
+  excludeRoute?: (pathname: string) => boolean;
+  /**
+   * Configuration for your Sitecore Experience Edge endpoint
+   */
+  edgeConfig: Omit<GraphQLPersonalizeServiceConfig, 'fetch'>;
+  /**
+   * Configuration for your Sitecore CDP endpoint
+   */
+  cdpConfig: Omit<CdpServiceConfig, 'dataFetcherResolver'>;
+};
+
+/**
+ * Middleware / handler to support Sitecore Personalize
+ */
+export class PersonalizeMiddleware {
+  private personalizeService: GraphQLPersonalizeService;
+  private cdpService: CdpService;
+
+  /**
+   * @param {PersonalizeMiddlewareConfig} [config] Personalize middleware config
+   */
+  constructor(protected config: PersonalizeMiddlewareConfig) {
+    // NOTE: we provide native fetch for compatibility on Next.js Edge Runtime
+    // (underlying default 'cross-fetch' is not currently compatible: https://github.com/lquixada/cross-fetch/issues/78)
+    this.personalizeService = new GraphQLPersonalizeService({
+      ...config.edgeConfig,
+      fetch: fetch,
+    });
+    // NOTE: same here, we provide NativeDataFetcher for compatibility on Next.js Edge Runtime
+    this.cdpService = new CdpService({
+      ...config.cdpConfig,
+      dataFetcherResolver: <T>() => {
+        const fetcher = new NativeDataFetcher({
+          debugger: debug.personalize,
+        });
+        return (url: string, data?: unknown) => fetcher.fetch<T>(url, data);
+      },
+    });
+  }
+
+  /**
+   * Gets the Next.js middleware handler
+   * @returns middleware handler
+   */
+  public getHandler(): (req: NextRequest, res?: NextResponse) => Promise<NextResponse> {
+    return this.handler;
+  }
+
+  protected get browserIdCookieName(): string {
+    // Each user should have saved identifier to connect between session, CDP uses bid cookies + local storage
+    return `bid_${this.config.cdpConfig.clientKey}`;
+  }
+
+  protected getBrowserId(req: NextRequest): string | undefined {
+    return req.cookies[this.browserIdCookieName] || undefined;
+  }
+
+  protected setBrowserId(res: NextResponse, browserId: string) {
+    if (browserId.length > 0) {
+      const expiryDate = new Date(new Date().setFullYear(new Date().getFullYear() + 2));
+      const options = { expires: expiryDate, secure: true };
+      res.cookie(this.browserIdCookieName, browserId, options);
+    }
+  }
+
+  private excludeRoute(pathname: string) {
+    if (
+      pathname.includes('.') || // Ignore files
+      pathname.startsWith('/api') // Ignore API calls
```

Comment:
It's possible that pathname could have a route name that starts with `/api` in this case it will be also excluded, need to do more strict check, using RegExp probably

Reply:
I updated to `/api/` which should reduce any accidental exclusions of legitimate routes. We provide the [`excludeRoute`](https://github.com/Sitecore/jss/pull/1008/files#diff-c5f8141e244ca57373cab027e38bab2cbf958cefb82c4fbbbfd7779493b8bdb3R19) option for them to customize further if necessary.